src/arch/cortex-m0/ctx_switch.S

```diff
@@ -27,29 +27,61 @@ EXC_RETURN_PSP:
   .word 0xFFFFFFFD
 
 .thumb_func
-/* r0 is top of user stack, r1 is heap base */
+/* r0 is top of user stack, r1 Process GOT */
 switch_to_user:
-  /* Load bottom of stack into Process Stack Pointer */
-  msr psp, r0
-
-  /* Cortex-M0 can only push registers R0-R7 directly, so move R8-R11 to R0-R3.
-   * This is equivalent to the 32-bit "push {r4-r11}" instruction. */
-  push {r4-r7}
-  mov r4,  r8
-  mov r5,  r9
-  mov r6, r10
-  mov r7, r11
-  push {r4-r7}
-
-  mov r9, r1
-  svc 0xff
-
-  /* These instructions are equivalent to the 32-bit "pop {r4-r11}" */
-  pop {r4-r7}
-  mov  r8, r4
-  mov  r9, r5
-  mov r10, r6
-  mov r11, r7
-  pop {r4-r7}
-
-  bx lr
+    /* Cortex-M0 can only push registers R0-R7 directly, so move R8-R11 to R0-R3.
+     * This is equivalent to the 32-bit "push {r4-r11}" instruction. */
+    push {r4-r7}
+    mov r4,  r8
+    mov r5,  r9
+    mov r6, r10
+    mov r7, r11
+    push {r4-r7}
+
+    /* Load non-hardware-stacked registers from Process stack */
+    ldmia r0!, {r4-r7}
+    mov r4,  r8
+    mov r5,  r9
+    mov r6, r10
+    mov r7, r11
+    ldmia r0!, {r4-r7}
+
+    /* Load bottom of stack into Process Stack Pointer */
+    msr psp, r0
+
+    /* Set PIC base pointer to the Process GOT */
+    mov r9, r1
+
+    /* SWITCH */
+    svc 0xff /* It doesn't matter which SVC number we use here */
+
+    /* Push non-harware-stacked registers onto Process stack.  r0 points to user
+     * stack (see to_kernel). We store the registers passed the stack to make it
+     * easier to address hardware-stacked registers in the rest of the kernel.
+     * This is fine since we just account for that above.
+     */
+    str r4, [r0, #32]
+    str r5, [r0, #28]
+    str r6, [r0, #24]
+    str r7, [r0, #20]
+
+    mov  r8, r4
+    mov  r9, r5
+    mov r10, r6
+    mov r11, r7
```

Comment:
Are these `mov`s opposite too? This code stores the same four registers twice currently.